docs/source/apidocs/castor_etc/castor_etc.parameters.rst
:py:mod:`castor_etc.parameters`
===============================

.. py:module:: castor_etc.parameters

.. autodoc2-docstring:: castor_etc.parameters
   :parser: myst
   :allowtitles:

Module Contents
---------------

Data
~~~~

.. list-table::
   :class: autosummary longtable
   :align: left

   * - :py:obj:`PASSBANDS <castor_etc.parameters.PASSBANDS>`
     - .. autodoc2-docstring:: castor_etc.parameters.PASSBANDS
          :parser: myst
          :summary:
   * - :py:obj:`PASSBAND_LIMITS <castor_etc.parameters.PASSBAND_LIMITS>`
     - .. autodoc2-docstring:: castor_etc.parameters.PASSBAND_LIMITS
          :parser: myst
          :summary:
   * - :py:obj:`PASSBAND_RESOLUTION <castor_etc.parameters.PASSBAND_RESOLUTION>`
     - .. autodoc2-docstring:: castor_etc.parameters.PASSBAND_RESOLUTION
          :parser: myst
          :summary:
   * - :py:obj:`PASSBAND_FILEPATHS <castor_etc.parameters.PASSBAND_FILEPATHS>`
     - .. autodoc2-docstring:: castor_etc.parameters.PASSBAND_FILEPATHS
          :parser: myst
          :summary:
   * - :py:obj:`PASSBAND_FILEUNITS <castor_etc.parameters.PASSBAND_FILEUNITS>`
     - .. autodoc2-docstring:: castor_etc.parameters.PASSBAND_FILEUNITS
          :parser: myst
          :summary:
   * - :py:obj:`PSF_FILEPATHS <castor_etc.parameters.PSF_FILEPATHS>`
     - .. autodoc2-docstring:: castor_etc.parameters.PSF_FILEPATHS
          :parser: myst
          :summary:
   * - :py:obj:`PSF_SUPERSAMPLE_FACTOR <castor_etc.parameters.PSF_SUPERSAMPLE_FACTOR>`
     - .. autodoc2-docstring:: castor_etc.parameters.PSF_SUPERSAMPLE_FACTOR
          :parser: myst
          :summary:
   * - :py:obj:`FWHM <castor_etc.parameters.FWHM>`
     - .. autodoc2-docstring:: castor_etc.parameters.FWHM
          :parser: myst
          :summary:
   * - :py:obj:`PX_SCALE <castor_etc.parameters.PX_SCALE>`
     - .. autodoc2-docstring:: castor_etc.parameters.PX_SCALE
          :parser: myst
          :summary:
   * - :py:obj:`PX_AREA <castor_etc.parameters.PX_AREA>`
     - .. autodoc2-docstring:: castor_etc.parameters.PX_AREA
          :parser: myst
          :summary:
   * - :py:obj:`IFOV_DIMEN <castor_etc.parameters.IFOV_DIMEN>`
     - .. autodoc2-docstring:: castor_etc.parameters.IFOV_DIMEN
          :parser: myst
          :summary:
   * - :py:obj:`IFOV_AREA <castor_etc.parameters.IFOV_AREA>`
     - .. autodoc2-docstring:: castor_etc.parameters.IFOV_AREA
          :parser: myst
          :summary:
   * - :py:obj:`TRANSIT_FOV <castor_etc.parameters.TRANSIT_FOV>`
     - .. autodoc2-docstring:: castor_etc.parameters.TRANSIT_FOV
          :parser: myst
          :summary:
   * - :py:obj:`MP <castor_etc.parameters.MP>`
     - .. autodoc2-docstring:: castor_etc.parameters.MP
          :parser: myst
          :summary:
   * - :py:obj:`TRANSIT_CCD_DIMENSIONS <castor_etc.parameters.TRANSIT_CCD_DIMENSIONS>`
     - .. autodoc2-docstring:: castor_etc.parameters.TRANSIT_CCD_DIMENSIONS
          :parser: myst
          :summary:
   * - :py:obj:`MIRROR_DIAMETER <castor_etc.parameters.MIRROR_DIAMETER>`
     - .. autodoc2-docstring:: castor_etc.parameters.MIRROR_DIAMETER
          :parser: myst
          :summary:
   * - :py:obj:`MIRROR_AREA <castor_etc.parameters.MIRROR_AREA>`
     - .. autodoc2-docstring:: castor_etc.parameters.MIRROR_AREA
          :parser: myst
          :summary:
   * - :py:obj:`DARK_CURRENT <castor_etc.parameters.DARK_CURRENT>`
     - .. autodoc2-docstring:: castor_etc.parameters.DARK_CURRENT
          :parser: myst
          :summary:
   * - :py:obj:`BIAS <castor_etc.parameters.BIAS>`
     - .. autodoc2-docstring:: castor_etc.parameters.BIAS
          :parser: myst
          :summary:
   * - :py:obj:`READ_NOISE <castor_etc.parameters.READ_NOISE>`
     - .. autodoc2-docstring:: castor_etc.parameters.READ_NOISE
          :parser: myst
          :summary:
   * - :py:obj:`GAIN <castor_etc.parameters.GAIN>`
     - .. autodoc2-docstring:: castor_etc.parameters.GAIN
          :parser: myst
          :summary:
   * - :py:obj:`REDLEAK_THRESHOLDS <castor_etc.parameters.REDLEAK_THRESHOLDS>`
     - .. autodoc2-docstring:: castor_etc.parameters.REDLEAK_THRESHOLDS
          :parser: myst
          :summary:
   * - :py:obj:`EXTINCTION_COEFFS <castor_etc.parameters.EXTINCTION_COEFFS>`
     - .. autodoc2-docstring:: castor_etc.parameters.EXTINCTION_COEFFS
          :parser: myst
          :summary:

API
~~~

.. py:data:: PASSBANDS
   :canonical: castor_etc.parameters.PASSBANDS
   :value: ['uv', 'u', 'g']

   .. autodoc2-docstring:: castor_etc.parameters.PASSBANDS
      :parser: myst

.. py:data:: PASSBAND_LIMITS
   :canonical: castor_etc.parameters.PASSBAND_LIMITS
   :value: None

   .. autodoc2-docstring:: castor_etc.parameters.PASSBAND_LIMITS
      :parser: myst

.. py:data:: PASSBAND_RESOLUTION
   :canonical: castor_etc.parameters.PASSBAND_RESOLUTION
   :value: None

   .. autodoc2-docstring:: castor_etc.parameters.PASSBAND_RESOLUTION
      :parser: myst

.. py:data:: PASSBAND_FILEPATHS
   :canonical: castor_etc.parameters.PASSBAND_FILEPATHS
   :value: None

   .. autodoc2-docstring:: castor_etc.parameters.PASSBAND_FILEPATHS
      :parser: myst

.. py:data:: PASSBAND_FILEUNITS
   :canonical: castor_etc.parameters.PASSBAND_FILEUNITS
   :value: None

   .. autodoc2-docstring:: castor_etc.parameters.PASSBAND_FILEUNITS
      :parser: myst

.. py:data:: PSF_FILEPATHS
   :canonical: castor_etc.parameters.PSF_FILEPATHS
   :value: None

   .. autodoc2-docstring:: castor_etc.parameters.PSF_FILEPATHS
      :parser: myst

.. py:data:: PSF_SUPERSAMPLE_FACTOR
   :canonical: castor_etc.parameters.PSF_SUPERSAMPLE_FACTOR
   :value: 20

   .. autodoc2-docstring:: castor_etc.parameters.PSF_SUPERSAMPLE_FACTOR
      :parser: myst

.. py:data:: FWHM
   :canonical: castor_etc.parameters.FWHM
   :value: None

   .. autodoc2-docstring:: castor_etc.parameters.FWHM
      :parser: myst

.. py:data:: PX_SCALE
   :canonical: castor_etc.parameters.PX_SCALE
   :value: None

   .. autodoc2-docstring:: castor_etc.parameters.PX_SCALE
      :parser: myst

.. py:data:: PX_AREA
   :canonical: castor_etc.parameters.PX_AREA
   :value: None

   .. autodoc2-docstring:: castor_etc.parameters.PX_AREA
      :parser: myst

.. py:data:: IFOV_DIMEN
   :canonical: castor_etc.parameters.IFOV_DIMEN
   :value: None

   .. autodoc2-docstring:: castor_etc.parameters.IFOV_DIMEN
      :parser: myst

.. py:data:: IFOV_AREA
   :canonical: castor_etc.parameters.IFOV_AREA
   :value: None

   .. autodoc2-docstring:: castor_etc.parameters.IFOV_AREA
      :parser: myst

.. py:data:: TRANSIT_FOV
   :canonical: castor_etc.parameters.TRANSIT_FOV
   :value: None

   .. autodoc2-docstring:: castor_etc.parameters.TRANSIT_FOV
      :parser: myst

.. py:data:: MP
   :canonical: castor_etc.parameters.MP
   :value: 930

   .. autodoc2-docstring:: castor_etc.parameters.MP
      :parser: myst

.. py:data:: TRANSIT_CCD_DIMENSIONS
   :canonical: castor_etc.parameters.TRANSIT_CCD_DIMENSIONS
   :value: [2048, 2048]

   .. autodoc2-docstring:: castor_etc.parameters.TRANSIT_CCD_DIMENSIONS
      :parser: myst

.. py:data:: MIRROR_DIAMETER
   :canonical: castor_etc.parameters.MIRROR_DIAMETER
   :value: None

   .. autodoc2-docstring:: castor_etc.parameters.MIRROR_DIAMETER
      :parser: myst

.. py:data:: MIRROR_AREA
   :canonical: castor_etc.parameters.MIRROR_AREA
   :value: None

   .. autodoc2-docstring:: castor_etc.parameters.MIRROR_AREA
      :parser: myst

.. py:data:: DARK_CURRENT
   :canonical: castor_etc.parameters.DARK_CURRENT
   :value: 0.0001

   .. autodoc2-docstring:: castor_etc.parameters.DARK_CURRENT
      :parser: myst

.. py:data:: BIAS
   :canonical: castor_etc.parameters.BIAS
   :value: 100

   .. autodoc2-docstring:: castor_etc.parameters.BIAS
      :parser: myst

.. py:data:: READ_NOISE
   :canonical: castor_etc.parameters.READ_NOISE
   :value: 3.0

   .. autodoc2-docstring:: castor_etc.parameters.READ_NOISE
      :parser: myst

.. py:data:: GAIN
   :canonical: castor_etc.parameters.GAIN
   :value: 2.0

   .. autodoc2-docstring:: castor_etc.parameters.GAIN
      :parser: myst

.. py:data:: REDLEAK_THRESHOLDS
   :canonical: castor_etc.parameters.REDLEAK_THRESHOLDS
   :value: None

   .. autodoc2-docstring:: castor_etc.parameters.REDLEAK_THRESHOLDS
      :parser: myst

.. py:data:: EXTINCTION_COEFFS
   :canonical: castor_etc.parameters.EXTINCTION_COEFFS
   :value: None

   .. autodoc2-docstring:: castor_etc.parameters.EXTINCTION_COEFFS
      :parser: myst


CHANGELOG.md
# Changelog

See [semantic versioning](https://semver.org/spec/v2.0.0.html) for the rationale behind
the version numbers.

## Current version (as of 2025-02-24)

- Fixed a bug ([#31](https://github.com/CASTOR-telescope/ETC/issues/31)) in the definition
  of `ellipticity`/`eccentricity` for a `GalaxySource`

## [1.3.2](https://github.com/CASTOR-telescope/ETC/tree/v1.3.2) (2024-11-10)

- Fixed the broken JupyterLab terminal on CANFAR
- Removed the `even` parameter in all `scipy.integrate.simpson()` calls, since this
  parameter is removed in the newest versions of scipy.

## [1.3.1](https://github.com/CASTOR-telescope/ETC/tree/v1.3.1) (2024-01-02)

- Added an `_encircled_energies` attribute to the `Photometry` object. This is needed for
  v1.1.1 of the [ETC frontend](https://github.com/CASTOR-telescope/ETC_frontend)
  ([#26](https://github.com/CASTOR-telescope/ETC/pull/26)).

## [1.3.0](https://github.com/CASTOR-telescope/ETC/tree/v1.3.0) (2023-11-13)

- Change `MP` and `READ_NOISE` parameters in parameters.py to 930 and 3.0, respectively.
  Also add some comments about parameters
  ([#22](https://github.com/CASTOR-telescope/ETC/pull/22))
- Introduced a new feature of Grism spectroscopy. This is very much still in development
  and currently only supports Sersic galaxy sources. Available results are: 2D dispersed
  spectrum and 1D SNR dispersed spectrum.
- Adapted transit numerical simulation from the POET mission. Using CASTOR tools, the
  light curve simulation calculation has been reduced from ~ 4 minutes to 1 second.
- Added a new feature that allows users to query GaiaDR2 database on the basis of
  (ra,dec), and threshold Gaia G magnitude, and select an appropriate normalized spectrum.
  This is achievable by selecting `Gaia` from the Predefined Spectra drop-down menu for a
  Point Source. To do transit simulation, one has to choose a Gaia Point Source.
- Added a new function `show_slit_image` in the uvmos_spectroscopy.py file to display a
  cropped view of the source on the detector.

## [1.2.1](https://github.com/CASTOR-telescope/ETC/tree/v1.2.1) (2023-09-18)

- Add ability to control the colour bar normalization when visualizing PSFs
  ([#16](https://github.com/CASTOR-telescope/ETC/pull/16))
- Update the bundled CASTOR PSFs ([#17](https://github.com/CASTOR-telescope/ETC/pull/17))

## [1.2.0](https://github.com/CASTOR-telescope/ETC/tree/v1.2.0) (2023-08-25)

- Changed AB magnitude calculation to interpolate the passband to the spectrum resolution
  ([#7](https://github.com/CASTOR-telescope/ETC/pull/7))
- Added PSF convolution to all sources using CASTOR PSFs
  ([#11](https://github.com/CASTOR-telescope/ETC/pull/11))
  - The fraction of flux enclosed within the aperture is estimated from the supersampled
    simulated images before binning down to the telescope's resolution. This works for any
    aperture with that is arbitrarily centred on the source. In this way, this change
    fixes the normalization limitation from v1.1.1, whose reference flux assumed galaxy
    sources were centred on and unrotated relative to the aperture
  - Now all photometry calculations use the fraction of flux enclosed within the aperture
    to calculate the electrons produced per second on the detector due to the source. Note
    that the reference flux, which determines what we consider to be 100% of the flux, is
    always based on the _noiseless_ image, i.e., with no PSF convolution
    - An enclosed flux fraction of 100% corresponds to the magnitude of the source that
      the user set. So if the user normalizes a spectrum to, say, an AB magnitude of 25,
      then this AB magnitude will be the AB magnitude of the source if 100% of its flux
      was contained within the aperture. If the user selects an aperture that only
      contains 50% of the flux, however, the effective AB magnitude will be dimmer
  - The photometry ETC is now 100% pixel-based, even for point sources. Changing any pixel
    in the `source_weights` attribute, which gives the fraction of the total flux
    contained within the pixel, _will_ change the photometry calculations

## [1.1.1](https://github.com/CASTOR-telescope/ETC/tree/v1.1.1) (2022-09-09)

- Jennifer Glover added UVMOS capabilities to ETC
- Tyrone Woods fixed a bug (noticed by Michael Balogh?) in `photometry.py` re: high Sérsic
  index galaxies

## [1.0.0](https://github.com/CASTOR-telescope/ETC/tree/v1.0.0) (2022-05-11)

Initial release!


castor_etc/parameters.py
#         GNU General Public License v3 (GNU GPLv3)
#
# (c) 2022.                            (c) 2022.
# Government of Canada                 Gouvernement du Canada
# National Research Council            Conseil national de recherches
# Ottawa, Canada, K1A 0R6              Ottawa, Canada, K1A 0R6
# All rights reserved                  Tous droits réservés
#
# NRC disclaims any warranties,        Le CNRC dénie toute garantie
# expressed, implied, or               énoncée, implicite ou légale,
# statutory, of any kind with          de quelque nature que ce
# respect to the software,             soit, concernant le logiciel,
# including without limitation         y compris sans restriction
# any warranty of merchantability      toute garantie de valeur
# or fitness for a particular          marchande ou de pertinence
# purpose. NRC shall not be            pour un usage particulier.
# liable in any event for any          Le CNRC ne pourra en aucun cas
# damages, whether direct or           être tenu responsable de tout
# indirect, special or general,        dommage, direct ou indirect,
# consequential or incidental,         particulier ou général,
# arising from the use of the          accessoire ou fortuit, résultant
# software. Neither the name           de l'utilisation du logiciel. Ni
# of the National Research             le nom du Conseil National de
# Council of Canada nor the            Recherches du Canada ni les noms
# names of its contributors may        de ses  participants ne peuvent
# be used to endorse or promote        être utilisés pour approuver ou
# products derived from this           promouvoir les produits dérivés
# software without specific prior      de ce logiciel sans autorisation
# written permission.                  préalable et particulière
#                                      par écrit.
#
# This file is part of the             Ce fichier fait partie du projet
# FORECASTOR ETC project.              FORECASTOR ETC.
#
# FORECASTOR ETC is free software:     FORECASTOR ETC est un logiciel
# you can redistribute it and/or       libre ; vous pouvez le redistribuer
# modify it under the terms of         ou le modifier suivant les termes de
# the GNU General Public               la "GNU General Public
# License as published by the          License" telle que publiée
# Free Software Foundation,            par la Free Software Foundation :
# either version 3 of the              soit la version 3 de cette
# License, or (at your option)         licence, soit (à votre gré)
# any later version.                   toute version ultérieure.
#
# FORECASTOR ETC is distributed        FORECASTOR ETC est distribué
# in the hope that it will be          dans l'espoir qu'il vous
# useful, but WITHOUT ANY WARRANTY;    sera utile, mais SANS AUCUNE
# without even the implied warranty    GARANTIE : sans même la garantie
# of MERCHANTABILITY or FITNESS FOR    implicite de COMMERCIALISABILITÉ
# A PARTICULAR PURPOSE. See the        ni d'ADÉQUATION À UN OBJECTIF
# GNU General Public License for       PARTICULIER. Consultez la Licence
# more details.                        Générale Publique GNU pour plus
#                                      de détails.
#
# You should have received             Vous devriez avoir reçu une
# a copy of the GNU General            copie de la Licence Générale
# Public License along with            Publique GNU avec FORECASTOR ETC ;
# FORECASTOR ETC. If not, see          si ce n'est pas le cas, consultez :
# <http://www.gnu.org/licenses/>.      <http://www.gnu.org/licenses/>.

"""
Parameters
==========

`castor_etc.parameters` contains predefined constatns for the CASTOR imaging chain.

"""

from math import pi
from os.path import join

import astropy.units as u

from .filepaths import DATAPATH

# -------------------------------- TELESCOPE PARAMETERS -------------------------------- #

# The telescope filter (i.e., passband) names
PASSBANDS = ["uv", "u", "g"]

# The passband wavelength limits for the different filters
PASSBAND_LIMITS = {
    "uv": [0.150, 0.300] << u.um,
    "u": [0.300, 0.400] << u.um,
    "g": [0.400, 0.550] << u.um,
}  # microns

# Resolution of the passband response curves
PASSBAND_RESOLUTION = 1 << u.nm  # nanometres

# Filepaths to the passband response curves
PASSBAND_FILEPATHS = {
    band: join(DATAPATH, "passbands", f"passband_castor.{band}") for band in PASSBANDS
}

# Wavelength units in the passband response curve files
PASSBAND_FILEUNITS = {"uv": u.um, "u": u.um, "g": u.um}

# Filepaths to the passband point spread functions (PSFs)
PSF_FILEPATHS = {
    band: join(DATAPATH, "psfs", f"{band}_psf_20x_supersampled.fits")
    for band in PASSBANDS
}

# The PSF oversampling factor. Each square pixel in the PSF file has a side length of
# PX_SCALE / PSF_SUPERSAMPLE_FACTOR (e.g., 0.1 arcsec / 20 = 0.005 arcsec)
PSF_SUPERSAMPLE_FACTOR = 20

# The PSF full-width at half-maximum
FWHM = 0.15 << u.arcsec  # arcsec

# The angular dimension covered by each pixel
PX_SCALE = 0.1 << u.arcsec  # arcsec (or arcsec/pixel)

# The angular area of each pixel
PX_AREA = PX_SCALE * PX_SCALE  # arcsec^2

# Instantaneous field of view
IFOV_DIMEN = [0.44, 0.56] << u.deg  # degrees. Angular dimensions
IFOV_AREA = IFOV_DIMEN[0] * IFOV_DIMEN[1]  # degrees^2

# Field of view of Poet
TRANSIT_FOV = 204.5 << u.arcsec

# Number of pixels in CCD (x 1 million)
MP = 930  # megapixels

# CCD dimensions of Poet
TRANSIT_CCD_DIMENSIONS = [2048,2048]  # pixels

# Aperture diameter
MIRROR_DIAMETER = 100 << u.cm  # cm

# Aperture area
MIRROR_AREA = pi * (0.5 * MIRROR_DIAMETER) * (0.5 * MIRROR_DIAMETER)  # cm^2

# (Teledyne e2v CMOS detector)
# Dark current is 0.01 electrons/s/pixel at -50°C and halves for every reduction of 5-6°C.
# CASTOR operates at 180 K, implying:
# dark current = 0.5^((223.15K - 180K) / 6) * 0.01 ~ 1e-4 electrons/s/pixel (negligible)
DARK_CURRENT = 1e-4  # electrons/s/pixel
# Dark current will increase linearly over time to 0.01 electrons/s/pixel by the end of 5
# years

BIAS = 100  # electron

READ_NOISE = 3.0  # electron/pixel (high-gain). Read noise is 30 electrons for low-gain

GAIN = 2.0  # electron/ADU

# Wavelength threshold for red leak. Flux longward of this is considered to be red leak
REDLEAK_THRESHOLDS = {"uv": 3010 << u.AA, "u": 4160 << u.AA, "g": 5600 << u.AA}

# Extinction coefficients (i.e., R := A/E(B-V)) for the different telescope passbands
# Estimates taken from Yuan+2013, column 3 of Table 2, rows: NUV, (SDSS) u, (SDSS) g
# <https://ui.adsabs.harvard.edu/abs/2013MNRAS.430.2188Y/abstract>
EXTINCTION_COEFFS = {"uv": 7.06, "u": 4.35, "g": 3.31}

# -------------------------------------------------------------------------------------- #
